inc/xmath.h

```diff
@@ -1449,116 +1471,214 @@ struct quat : public vec<4, QS>
+                         const vec<3>& ray_d,
+                         const vec<3>& sphere_o,
+                         const XMTYPE  sphere_r)
+{
+	auto l  = sphere_o - ray_o;
+	auto s  = l.dot(ray_d);
+	auto l2 = l.dot(l);
+	auto r2 = sphere_r * sphere_r;
 
-		for (unsigned i = 0; i < 3; i++)
-		{
-			auto e = p.dot(box_sides[i] / half_lengths[i]);
-			auto f = ray_d.dot(box_sides[i] / half_lengths[i]);
+	if (s < 0 && l2 > r2) { return NAN; }
 
-			if (static_cast<XMTYPE>(fabs(f)) > epsilon)
-			{
-				auto t_1 = (e + half_lengths[i]) / f;
-				auto t_2 = (e - half_lengths[i]) / f;
-
-				if (t_1 > t_2) { std::swap(t_1, t_2); }
-				if (t_1 > t_min) { t_min = t_1; }
-				if (t_2 < t_max) { t_max = t_2; }
-				if (t_min > t_max) { return NAN; }
-				if (t_max < 0) { return NAN; }
-			}
-			else if ((-e - half_lengths[i]) > 0 || (-e + half_lengths[i]) < 0)
-			{
-				return NAN;
-			}
-		}
+	auto m2 = l2 - (s * s);
 
-		if (t_min > 0) { return t_min; }
+	if (m2 > r2) { return NAN; }
 
-		return t_max;
-	}
+	auto q = sqrt(r2 - m2);
+
+	auto b = l2 > r2;
+	return b * (s - q) + !b * (s + q);
+}
 
-	static XMTYPE ray_sphere(const vec<3>& ray_o,
-	                         const vec<3>& ray_d,
-	                         const vec<3>& sphere_o,
-	                         const XMTYPE  sphere_r)
+}; // namespace intersect
+
+namespace filter
+{
+
+/**
+ * @brief      { struct_description }
+ *
+ * @tparam     Z     { description }
+ * @tparam     X     { description }
+ * @tparam     U     { description }
+ * @tparam     S     { description }
+ */
+template <size_t Z, size_t X, size_t U, typename S = XMTYPE>
+struct kalman
+{
+	struct
 	{
-		auto l  = sphere_o - ray_o;
-		auto s  = l.dot(ray_d);
-		auto l2 = l.dot(l);
-		auto r2 = sphere_r * sphere_r;
+		mat<X, 1, S> state;
+		mat<X, X, S> covariance = mat<X, X, S>::I();
+		mat<Z, 1, S> measurement_residual;
+	} estimated;
 
-		if (s < 0 && l2 > r2) { return NAN; }
+	mat<X, X, S> I_xx = mat<X, X, S>::I();
 
-		auto m2 = l2 - (s * s);
+	kalman(vec<X, S>& state = {}) { estimated.state = {state}; }
 
-		if (m2 > r2) { return NAN; }
+	/**
+	 * @brief      { function_description }
+	 *
+	 * @param[in]  state_transition  The state transition
+	 * @param[in]  control_to_state  The control to state
+	 * @param[in]  control           The control
+	 */
+	void time_update(const mat<X, X, S>& state_transition,
+	                 const mat<X, U, S>& control_to_state,
+	                 const mat<U, 1, S>& control,
+	                 const mat<X, X, S>& process_noise_covariance = {})
+	{
+		// Advance state in time by state_transition matrix. Apply control
+		// vector transformed into state space to state.
+		estimated.state = state_transition * estimated.state;
+		estimated.state += control_to_state * control;
 
-		auto q = sqrt(r2 - m2);
+		const auto& F = state_transition;
+		auto&       P = estimated.covariance;
+
+		// Inflate covariance. TODO: study
+		P = F * P * F.transpose() + process_noise_covariance;
```

Comment:
If the process noise covariance depends on the time between measurements, then you can use this algorithm: https://nyxspace.com/MathSpec/orbit_determination/snc/ .

If the process noise is static, then this works good.

Reply:
Isn't this the same thing? The `process_noise_covariance` is a parameter provided with the call to this method. The user may update it between calls as they see fit, so they could adjust it if the application is time dependent.